tasmota/tasmota_xdrv_driver/xdrv_92_esp32_gnss.ino

```diff
@@ -0,0 +1,876 @@
+{
+  if (gps.time.isValid() && gps.date.isValid() &&
+      ((current_time - last_rtc_sync) > RTC_UPDATE_INTERVAL))
+  {
+    TIME_T tm;
+    tm.year = gps.date.year() - 1970;
+    tm.month = gps.date.month();
+    tm.day_of_month = gps.date.day();
+    tm.hour = gps.time.hour();
+    tm.minute = gps.time.minute();
+    tm.second = gps.time.second();
+    tm.nanos = gps.time.centisecond() * 10000;
+
+    uint32_t epoch = MakeTime(tm);
+    RtcSetTime(epoch);
+    AddLog(LOG_LEVEL_INFO, PSTR("GNSS: RTC Time sync %s"), GNSSData.time);
+    last_rtc_sync = current_time;
+  }
+}
+
+static void UpdateLastValidTime(uint32_t current_time)
+{
+  if (GNSSData.last_valid > 0)
+  {
+    TIME_T last_valid_time_t;
+    BreakTime(current_time - GNSSData.last_valid, last_valid_time_t);
+    snprintf_P(GNSSData.last_valid_time, sizeof(GNSSData.last_valid_time),
+               PSTR("%dT%02d:%02d:%02d"),
+               last_valid_time_t.days, last_valid_time_t.hour,
+               last_valid_time_t.minute, last_valid_time_t.second);
+  }
+  else
+  {
+    strlcpy(GNSSData.last_valid_time, PSTR("N/A"), sizeof(GNSSData.last_valid_time));
+  }
+}
+
+static void ProcessGNSSMessage(const GNSSParser::Message &msg, uint32_t current_time)
+{
+  switch (msg.type)
+  {
+  case GNSSParser::Message::Type::NMEA:
+  {
+    if (ProcessNMEAMessage((char *)msg.data, msg.length))
+    {
+      UpdateGNSSData(current_time);
+    }
+    break;
+  }
+#ifdef USE_NTRIP
+  case GNSSParser::Message::Type::RTCM3:
+    ProcessRTCMMessage(msg.data, msg.length);
+    break;
+#endif
+  default:
+    break;
+  }
+}
+
+static void LogRawBuffer(const uint8_t *buffer, size_t length)
+{
+  if (!length)
+    return;
+
+  const size_t BYTES_PER_LINE = 16;
+  char hex_line[BYTES_PER_LINE * 3 + 1]; // 2 chars per byte + space + null
+  char ascii_line[BYTES_PER_LINE + 1];   // 1 char per byte + null
+
+  for (size_t offset = 0; offset < length; offset += BYTES_PER_LINE)
+  {
+    size_t line_length = std::min(BYTES_PER_LINE, length - offset);
+
+    memset(hex_line, ' ', sizeof(hex_line));
+    memset(ascii_line, ' ', sizeof(ascii_line));
+    hex_line[sizeof(hex_line) - 1] = 0;
+    ascii_line[sizeof(ascii_line) - 1] = 0;
+
+    for (size_t i = 0; i < line_length; i++)
+    {
+      uint8_t byte = buffer[offset + i];
+      snprintf(&hex_line[i * 3], 4, "%02X ", byte);
+      ascii_line[i] = (byte >= 32 && byte <= 126) ? byte : '.';
+    }
+
+    AddLog(LOG_LEVEL_DEBUG_MORE, PSTR("GNSS Raw [%04X]: %-48s  |%s|"),
+           offset, hex_line, ascii_line);
+  }
+}
+
+/*********************************************************************/
+/*             TASMOTA WEBSERVER DISPLAY (USE_WEBSERVER)             */
+/*********************************************************************/
+#ifdef USE_WEBSERVER
+void GNSSShow(bool json)
+{
+  if (json)
+  {
+    ResponseAppend_P(
+        PSTR(",\"GNSS\":{\"Valid\":%d,\"DateTime\":\"%s %s\",\"Lat\":%f,\"Lon\":%f,\"Alt\":%f,"
```

Comment:
You can make it work for ESP8266 as well, but need to use the special extensions

https://tasmota.github.io/docs/Sensor-API/#extension-to-vsnprintf 

It may also be useful to limit and control the number digits in the output